httpdocs/ajax/proto/create.php

```diff
@@ -0,0 +1,244 @@
+  if ($line[0] === '#') {
+      $line = trim(str_replace('#', '', $line));
+      if (strtolower(substr($line, 0, 8)) !== 'template' && strtolower(substr($line, 0, 17)) !== 'documentation url' && substr($line, 0, 4) !== 'VRML') {
+        if(strtolower(substr($line, 0, 4)) !== 'tags') {
+          if (strpos($line, 'deprecated') || strpos($line, 'hidden'))
+            error("This proto is either deprecated or hidden and should not be added.");
+        elseif (strtolower(substr($line, 0, 11)) === 'license url')
+          $license_url = trim(preg_replace("/license url\s*:/", '', $line));
+        elseif (strtolower(substr($line, 0, 7)) === 'license')
+          $license = trim(preg_replace("/license\s*:/", '', $line));
+        else {
+          if ($description !== '')
+            $description .= "\n";
+          $description .= $mysqli->escape_string($line);
+        }
+      }
+    }
+  } elseif (substr($line, 0, 11) === 'EXTERNPROTO') {
+    $proto_url = trim(str_replace('"', '',str_replace('EXTERNPROTO', '', $line)));
+    $proto_name = str_replace('.proto"', '', $line);
+    $proto_name = substr($proto_name, strrpos($proto_name, '/') + 1);
+    array_push($externprotos, [$proto_name, $proto_url]);
+  } elseif (substr($line, 0, 6) === 'PROTO ')
+    $title = trim(substr($line, 6));
+  if (!empty($title) && $title[-1] === '[')
+    $title = trim(substr($title, 0, -1));
+  $line = strtok("\r\n");
+}
+
+$original_externprotos = $externprotos;
+$title = $mysqli->escape_string($title);
+$license = $mysqli->escape_string($license);
+$license_url = $mysqli->escape_string($license_url);
+
+$base_type = '';
+preg_match("/(?:\]\s*)\{\s*(?:\%\<[\s\S]*?(?:\>\%\s*))?(?:DEF\s+[^\s]+)?\s+([a-zA-Z0-9\_\-\+]+)\s*\{/", $proto_content, $match);
+if ($match)
+  $base_type = $match[1];
+$base_proto = $base_type;
+$base_nodes = ['Gyro', 'DistanceSensor', 'Recognition', 'TouchSensor', 'ContactProperties', 'TextureCoordinate', 'Color',
+  'Camera', 'Accelerometer', 'Slot', 'Radar', 'Transform', 'Zoom', 'RangeFinder', 'PointSet', 'Capsule', 'Speaker', 'Lens',
+  'Viewpoint', 'IndexedFaceSet', 'Solid', 'Group', 'Muscle', 'Lidar', 'InertialUnit', 'DirectionalLight',
+  'HingeJointParameters', 'Compass', 'Normal', 'Propeller', 'Physics', 'RotationalMotor', 'Microphone', 'ImageTexture', 'Fog',
+  'Mesh', 'Track', 'Background', 'LED', 'Material', 'Box', 'PointLight', 'Cylinder', 'Damping', 'GPS', 'Radio', 'Pen', 'Cone',
+  'WorldInfo', 'SpotLight', 'TextureTransform', 'LinearMotor', 'Receiver', 'Coordinate', 'Hinge2JointParameters', 'CadShape',
+  'LensFlare', 'TrackWheel', 'PBRAppearance', 'Shape', 'Altimeter', 'PositionSensor', 'Connector', 'HingeJoint', 'Plane',
+  'Brake', 'Appearance', 'ElevationGrid', 'BallJointParameters', 'Fluid', 'Robot', 'SolidReference', 'Sphere', 'Skin',
+  'IndexedLineSet', 'ImmersionProperties', 'JointParameters', 'Focus', 'SliderJoint', 'Emitter', 'Hinge2Joint', 'BallJoint',
+  'LightSensor', 'Display', 'Billboard', 'Charger'];
```

Comment:
That's another difficult to maintain list... Should be at least added in the add new node developers' wiki.